File: chapter02/Sequence.rb
class Sequence < Struct.new(:first, :second)
  def reduce(environment)
    case first
    when DoNothing.new
      [second, environment]
    else
      reduce_first, reduced_environment = first.reduce(environment)
      [Sequence.new(reduce_first, second), reduced_environment]
    end
  end

  def reducible?
    true
  end

  def to_s
    "#{first}; #{second}"
  end

  def inspect
    "#{self}"
  end
end

File: chapter02/DenotationSemantic/Sequence.py
class Sequence(object):
    def __init__(self, first, second):
        self.first = first
        self.second = second

    def to_python(self):
        first = self.first.to_python()
        second = self.second.to_python()
        return '{first}\n{second}'.format(**locals())
